README.md

```diff
@@ -1 +1,115 @@
+
+0. this doc
+1. create the library for the process runner and basic authz.
+2. basic TLS-secured gRPC client-server chat. The API will wrapper the process runner library
+3. support for hard coded gRPC streaming data. Something simple like a clock tick
+4. Spawn some basic jobs, capture stderr/stdout and pump them into the streams
+5. Work on job control. Cancelling a process, ensuring the channels/goroutines/resources clean up correctly, etc. 
+6. Add cgroups and resource isolation 
+
+(this may change)
+
+Stretch goals
+
+7. some key tests
+8. docker packaging
+9. a nice client
+10. magic
+
+## API
+
+Exposed gRPC "Runner" service methods:
+```
+Auth(username) (Session, error)
+Run(Session, cmd, args...) (Process, error)
+Stdout(Session, Process) (Stream, error)
+Status(Session, Process) (Status, error)
+Abort(Session, Process) (Status, error)
+```
+
+where: 
+ * the `Auth` method returns a `Session` method to authenticated callers.
+ * the `Run` method that can launch new processes for authenticated callers.
+   * the "role" of the authenticated user will determine the cgroup the process runs in
+   * `Run` takes `cmd` and `args` args for the intended linux command and related cmdline parameters and returns a `Process` message
+ * `Process` is the process ID for the process control calls: `Stdout`, `Status`, and `Abort`
+ * `Status` will return an enum of `COMPLETED|RUNNING|ERROR` and, ideally, the exit code
+
+## AuthN / AuthZ
+
+A hardcoded set of usernames (no passwords) will be mapped to a set of V2 cgroups 
+
+| cgroup name | CPU | Memory | 
+| ------------| --- | ------ |
+| **good**    | x   | x      |
+| **better**  | x   | x      |
+| **best**    | x   | x      |
+
+| username | assigned cgroup | 
+| -------- | --------------- |
+| **good** | good            |
+| **better** | better        |
+| **best** | best            |
+
+The caller can log in with the `Auth()` method using one of the whitelisted usernames. Other usernames will result in an error. 
+A username may be used across multiple sessions. A successful login will return a `SessionID` to the caller which must be used for all subsequent operations. 
+
+The Session ID is just a stringified UUID. 
+
+## Client UX
+
+The client initially will just be a golang program that runs the server through some paces (just an integration test harness). But in later releases I will try to turn this in to a simple cmdline tool with basic support for each operation. Exact syntax TBD but it may look something like:
+
+```
+> client auth sandy
+xxx-yyyy
+> client -session xxx-yyy run ls -laR
+12345
+> client stdout -session xxx-yyy -process 12345
+.
+..
+README.md
+> client status -session xxx-yyy -process 12345
+COMPLETED
+Exit code: 0
+> client abort -session xxx-yyy -process 12345
+already terminated
+```
+
+Beyond the raw client code generated by the gRPC code there will not be a separate "sugar" client library. Although it would be handy for managing the stdout streams. 
```

Comment:
I'm not exactly sure what you mean by this, but remember the _library_ itself should support streaming, and then the _server_ wraps this in a gRPC stream. This means the library should have a convenient streaming API that is not coupled to gRPC and could be used in other contexts as well.

Reply:
Correct. I was wondering if there should be a similar library for the client? Or are you envisioning the library being used for both client and server?

Reply:
Gotcha. There's no need for a separate client library, since the server is exclusively gRPC its fine for the client to make gRPC calls directly as you suggest here. Thanks for clarifying!